### PR TITLE
Ignore conflicts in certain packages

### PR DESCRIPTION
**Pull request by iRevive**
This PR partially covers https://github.com/scalacenter/sbt-missinglink/issues/12.

Conflicts can be ignored based on the package name of the class that has the conflict.
Basically, this is just a backport from the maven plugin.

SBT:
```scala
missinglinkIgnoreDestinationPackages += IgnoredPackage("com.google.common", ignoreSubpackages = false)
missinglinkIgnoreSourcePackages += IgnoredPackage("com.example")
```

Maven:
```xml
<configuration>
  <!-- ignore conflicts with groovy.lang on the caller side -->
  <ignoreSourcePackages>
    <ignoreSourcePackages>
      <package>com.example</package>
    </ignoreSourcePackages>
    <ignoreSubpackages>false</ignoreSubpackages>
  </ignoreSourcePackages>
  <!-- ignore conflicts with com.foo on the callee side -->
  <ignoreDestinationPackages>
    <ignoreDestinationPackage>
      <package>com.google.common</package>
    </ignoreDestinationPackage>
  </ignoreDestinationPackages>
</configuration>
```

Also, this PR brings a new setting `missinglinkFailOnConflicts`:

SBT:
```scala
missinglinkFailOnConflicts := false
```

Maven:
```xml
<configuration>
  <failOnConflicts>true</failOnConflicts>
</configuration>
```

### CODE REVIEW

**Comment by iRevive** [comment]
Hi @sjrd. Once you have time, can you have a look, please?

File: src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala [hunk 1]
```diff
@@ -20,9 +20,29 @@ import com.spotify.missinglink.datamodel.{
...
     val missinglinkCheck: TaskKey[Unit] =
-      taskKey[Unit]("run the missinglink checks")
+      taskKey[Unit]("Run the missinglink checks")
+
+    val missinglinkFailOnConflicts: SettingKey[Boolean] =
+      settingKey[Boolean]("Fail the build if any conflicts are found")
+
+    val missinglinkIgnoreSourcePackages: SettingKey[Seq[IgnoredPackage]] =
+      settingKey[Seq[IgnoredPackage]](
+        "Optional list of packages to ignore conflicts in where the source of the conflict " +
```
Comment (sjrd) [inline src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala]:
```suggestion
        "Optional list of packages to ignore conflicts where the source of the conflict " +
```

File: src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala [hunk 2]
```diff
@@ -20,9 +20,29 @@ import com.spotify.missinglink.datamodel.{
...
+          "is in one of the specified packages."
+      )
+
+    val missinglinkIgnoreDestinationPackages: SettingKey[Seq[IgnoredPackage]] =
+      settingKey[Seq[IgnoredPackage]](
+        "Optional list of packages to ignore conflicts in where the destination/called-side " +
```
Comment (sjrd) [inline src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala]:
```suggestion
        "Optional list of packages to ignore conflicts where the destination/called-side " +
```

File: src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala [hunk 3]
```diff
@@ -20,9 +20,29 @@ import com.spotify.missinglink.datamodel.{
 }
 
 object MissingLinkPlugin extends AutoPlugin {
-  object autoImport {
+
+  trait Types {
+    final case class IgnoredPackage(name: String, ignoreSubpackages: Boolean = true)
```
Comment (sjrd) [inline src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala]: Please put this directly inside `autoImport`. There is no need to create a path-dependent case class inside a `trait Types`.

File: src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala [hunk 4]
```diff
@@ -36,14 +56,41 @@ object MissingLinkPlugin extends AutoPlugin {
...
-      outputConflicts(conflicts, log)
-      if (conflicts.nonEmpty)
-        throw new MessageOnlyException("there were conflicts")
+        val diffMessage = if (initialTotal != filteredTotal) {
+          s"($initialTotal conflicts were found before applying filters)"
+        } else {
+          ""
+        }
+
+        log.warn(s"$filteredTotal conflicts found! $diffMessage")
```
Comment (sjrd) [inline src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala]: This warning might get annoying. If one specifically ignored some conflicts, presumably they *don't* want to be reminded about them all the time. Consider using `log.info` instead.

File: src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala [hunk 5]
```diff
@@ -140,6 +187,60 @@ object MissingLinkPlugin extends AutoPlugin {
...
+    ignoreDestinationPackages: Seq[IgnoredPackage],
+    log: Logger
+  ): Seq[Conflict] = {
+
+    def filter(
+      ignoredPackages: Seq[IgnoredPackage],
+      name: String,
+      setting: String,
+      field: Dependency => ClassTypeDescriptor
+    ): Seq[Conflict] => Seq[Conflict] = input => {
```
Comment (sjrd) [inline src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala]:
```suggestion
    ): Seq[Conflict] => Seq[Conflict] = { input =>
```

File: src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala [hunk 6]
```diff
@@ -140,6 +187,60 @@ object MissingLinkPlugin extends AutoPlugin {
...
+            conflictPackageName == p.name ||
+            (p.ignoreSubpackages && conflictPackageName.startsWith(p.name + "."))
+          }
+        }
+
+        val filtered = input.filterNot(isIgnored)
+        val diff = input.length - filtered.length
+
+        if (diff != 0) {
+          log.warn(
```
Comment (sjrd) [inline src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala]: Same comment here. Consider using `log.info` instead.

File: src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala [hunk 7]
```diff
@@ -36,14 +56,41 @@ object MissingLinkPlugin extends AutoPlugin {
...
+        } else {
+          ""
+        }
+
+        log.warn(s"$filteredTotal conflicts found! $diffMessage")
+
+        outputConflicts(filteredConflicts, log)
+
+        if (failOnConflicts)
+          throw new MessageOnlyException(s"There were ($filteredTotal) conflicts")
```
Comment (sjrd) [inline src/main/scala/ch/epfl/scala/sbtmissinglink/MissingLinkPlugin.scala]: I'm not sure why the parens? Consider:
```suggestion
          throw new MessageOnlyException(s"There were $filteredTotal conflicts")
```

File: src/sbt-test/missinglink/do-not-fail-on-conflicts/build.sbt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+inThisBuild(Def.settings(
+  version := "0.1.0",
+  scalaVersion := "2.12.8",
+))
+
+lazy val `ignore-source-package` = project
```
Comment (sjrd) [inline src/sbt-test/missinglink/do-not-fail-on-conflicts/build.sbt]:
```suggestion
lazy val `do-not-fail-on-conflicts` = project
```
to match the test name. (I suspect a copy-paste error.)

**Comment by iRevive** [comment]
Hey! Thanks for the review. I will update the PR soon.

**Pull request merged**